
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>David Winter</title>
  <meta name="author" content="David Winter">

  
  <meta name="description" content="A couple of people have contacted me in the last little while to ask whether mmod,
my little R library for calculating population differentiation &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://dwinter.github.com/">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="/javascripts/ender.js"></script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <link href="/atom.xml" rel="alternate" title="David Winter" type="application/atom+xml">
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">

  

</head>

<body   >
  <header role="banner"><hgroup>
  <h1><a href="/">David Winter</a></h1>
  
    <h2>Code for evolutionary biology</h2>
  
</hgroup>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
  <li><a href="/atom.xml" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li>
  
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:dwinter.github.com" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div class="blog-index">
  
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2012/11/17/diff-stats-for-sequence-data/">MMOD Does Sequences</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2012-11-17T17:16:00+13:00" pubdate data-updated="true">Nov 17<span>th</span>, 2012</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>A couple of people have contacted me in the last little while to ask whether <code>mmod</code>,
my little R library for calculating population differentiation statistics, can be
used to deal with sequence data.</p>

<p>When I was designing the pacakge I was really thinking about handling microsattelite genotypes, but the statistics that mmod calculates are actually all capable of deally with almost any genetic data. Although the actualy calculations can be handled easily enough, it&#8217;s a bit of a pain coverting standard sequence datasets into population
genetic ones. So, I&#8217;ve written a little function to aid the conversion of the standard representation of DNA sequences in R (<code>DNAbin</code> objects) into the <code>genind</code>
object that is used to store population genetic data.</p>

<p>I&#8217;ll use this post to show you how it works, and to compare the &#8220;identity-based&#8221;
statisitics mmod caluclates with the standard AMOVA approach to population
differentiation from sequence data.</p>

<h2>AMOVA</h2>

<p>The most widely used method of measuring population differentiation from sequence
data is  Excoffier&#8217;s AMOVA (Analysis of Molecular Variance). This method takes into
account the relationship betwen haplotypes in terms of genetic distance between them
and paritions variation in those distances into within- and among-population components. It also produces an Fst-like statisc for every level of population strucure being considered.</p>

<p><a href="http://ape.mpl.ird.fr/ep/">Emmanuel Paradis</a>&#8217;  library <code>pegas</code> can calulate
AMOVA for a given dataset. Here&#8217;s an example use some faked-up data:</p>

<figure class='code'><figcaption><span></span></figcaption><div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
</pre></td><td class='code'><pre><code class='r'><span class='line'>library<span class="p">(</span>pegas<span class="p">)</span>
</span><span class='line'>data<span class="p">(</span>woodmouse<span class="p">)</span>
</span><span class='line'>set.seed<span class="p">(</span><span class="m">123</span><span class="p">)</span>
</span><span class='line'><span class="p">(</span>fake_data <span class="o">&lt;-</span> woodmouse<span class="p">[</span>sample<span class="p">(</span><span class="m">1</span>:<span class="m">15</span><span class="p">,</span> <span class="m">100</span><span class="p">,</span> replace <span class="o">=</span> <span class="kc">TRUE</span><span class="p">),</span> <span class="p">])</span>
</span></code></pre></td></tr></table></div></figure>




<figure class='code'><figcaption><span></span></figcaption><div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
</pre></td><td class='code'><pre><code class='r'><span class='line'><span class="c1">## 100 DNA sequences in binary format stored in a matrix.</span>
</span><span class='line'><span class="c1">## </span>
</span><span class='line'><span class="c1">## All sequences of same length: 965 </span>
</span><span class='line'><span class="c1">## </span>
</span><span class='line'><span class="c1">## Labels: No0908S No1114S No0910S No1206S No1208S No305 ...</span>
</span><span class='line'><span class="c1">## </span>
</span><span class='line'><span class="c1">## Base composition:</span>
</span><span class='line'><span class="c1">##     a     c     g     t </span>
</span><span class='line'><span class="c1">## 0.306 0.261 0.126 0.307 </span>
</span></code></pre></td></tr></table></div></figure>


<p>To measure population differentiation we obviously need some populations, the
<code>amova</code> function takes a vector assiging each individual in the dataset to a
particular location (note, you aren&#8217;t limited to two levels of population structure,
you can use R&#8217;s  formula syntax to denote, say, ~samples/sites/regions):</p>

<figure class='code'><figcaption><span></span></figcaption><div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
</pre></td><td class='code'><pre><code class='r'><span class='line'>pops <span class="o">&lt;-</span> factor<span class="p">(</span>rep<span class="p">(</span>c<span class="p">(</span><span class="s">&quot;Dunedin&quot;</span><span class="p">,</span> <span class="s">&quot;Mosgeil&quot;</span><span class="p">),</span> each <span class="o">=</span> <span class="m">50</span><span class="p">))</span>
</span><span class='line'>dm <span class="o">&lt;-</span> dist.dna<span class="p">(</span>fake_data<span class="p">,</span> <span class="s">&quot;N&quot;</span><span class="p">)</span>
</span><span class='line'><span class="p">(</span>res <span class="o">&lt;-</span> amova<span class="p">(</span>dm ~ pops<span class="p">))</span>
</span></code></pre></td></tr></table></div></figure>




<figure class='code'><figcaption><span></span></figcaption><div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
</pre></td><td class='code'><pre><code class='r'><span class='line'><span class="c1">## </span>
</span><span class='line'><span class="c1">##   Analysis of Molecular Variance</span>
</span><span class='line'><span class="c1">## </span>
</span><span class='line'><span class="c1">## Call: amova(formula = dm ~ pops)</span>
</span><span class='line'><span class="c1">## </span>
</span><span class='line'><span class="c1">##           SSD   MSD df</span>
</span><span class='line'><span class="c1">## pops    41.72 41.72  1</span>
</span><span class='line'><span class="c1">## Error 7449.26 76.01 98</span>
</span><span class='line'><span class="c1">## Total 7490.98 75.67 99</span>
</span><span class='line'><span class="c1">## </span>
</span><span class='line'><span class="c1">## Variance components:</span>
</span><span class='line'><span class="c1">##       sigma2 P.value</span>
</span><span class='line'><span class="c1">## pops  -0.686    0.71</span>
</span><span class='line'><span class="c1">## Error 76.013        </span>
</span><span class='line'><span class="c1">## </span>
</span><span class='line'><span class="c1">## Variance coefficients:</span>
</span><span class='line'><span class="c1">##  a </span>
</span><span class='line'><span class="c1">## 50 </span>
</span><span class='line'><span class="c1">## </span>
</span></code></pre></td></tr></table></div></figure>


<p>For what ever reason the <code>amova</code> doesn&#8217;t give the Fst like stats, but we can
do that for ourselves:</p>

<figure class='code'><figcaption><span></span></figcaption><div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='r'><span class='line'>sigma2 <span class="o">&lt;-</span> res<span class="p">$</span>varcomp<span class="p">[</span><span class="m">1</span><span class="p">,</span> <span class="m">1</span><span class="p">]</span>
</span><span class='line'><span class="p">(</span>phi_st <span class="o">&lt;-</span> <span class="p">(</span>sigma2<span class="o">/</span><span class="p">(</span>sigma2 <span class="o">-</span> res<span class="p">$</span>tab<span class="p">[</span><span class="m">1</span><span class="p">,</span> <span class="m">2</span><span class="p">])))</span>
</span></code></pre></td></tr></table></div></figure>


<figure class='code'><figcaption><span></span></figcaption><div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='r'><span class='line'><span class="c1">##[1] 0.0161</span>
</span></code></pre></td></tr></table></div></figure>


<h2>Everthing else (mmod)</h2>

<p>AMOVA is very useful, but, like G_st it can be difficult to compare between
studies because the Phi statistics are partially determined by the total diversity
in the dataset. There is not obvious way to correct for this effect.</p>

<p>An alternative approach is to discard the information about how haplotypes relate
to each other and simply use the frequency of each haplotype present to calculate
differentiation statistics from allele frequencies, in the same way you would
 for SNPs or microsattelites. The latest build of <code>mmod</code> has a function, <code>as.genind.DNAbin</code> to make this easy. It&#8217;s not on CRAN just yet, but you can
 install it from github. The function takes a <code>DNAbin</code> object and a vector
 of populations for each sequence:</p>

<figure class='code'><figcaption><span></span></figcaption><div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
</pre></td><td class='code'><pre><code class='r'><span class='line'>library<span class="p">(</span>devtools<span class="p">)</span>
</span><span class='line'>install_github<span class="p">(</span>rep<span class="o">=</span><span class="s">&#39;mmod&#39;</span><span class="p">,</span> username<span class="o">=</span><span class="s">&#39;dwinter&#39;</span><span class="p">)</span>
</span><span class='line'>library<span class="p">(</span>mmod<span class="p">)</span>
</span><span class='line'><span class="p">(</span>fake_genind <span class="o">&lt;-</span> as.genind.DNAbin<span class="p">(</span>x<span class="o">=</span>fake_data<span class="p">,</span> pop<span class="o">=</span>pops<span class="p">))</span>
</span></code></pre></td></tr></table></div></figure>




<figure class='code'><figcaption><span></span></figcaption><div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
<span class='line-number'>17</span>
<span class='line-number'>18</span>
<span class='line-number'>19</span>
<span class='line-number'>20</span>
<span class='line-number'>21</span>
<span class='line-number'>22</span>
<span class='line-number'>23</span>
<span class='line-number'>24</span>
<span class='line-number'>25</span>
</pre></td><td class='code'><pre><code class='r'><span class='line'><span class="c1">## </span>
</span><span class='line'><span class="c1">##    #####################</span>
</span><span class='line'><span class="c1">##    ### Genind object ### </span>
</span><span class='line'><span class="c1">##    #####################</span>
</span><span class='line'><span class="c1">## - genotypes of individuals - </span>
</span><span class='line'><span class="c1">## </span>
</span><span class='line'><span class="c1">## S4 class:  genind</span>
</span><span class='line'><span class="c1">## @call: genind(tab = tab, pop = pops)</span>
</span><span class='line'><span class="c1">## </span>
</span><span class='line'><span class="c1">## @tab:  100 x 15 matrix of genotypes</span>
</span><span class='line'><span class="c1">## </span>
</span><span class='line'><span class="c1">## @ind.names: vector of  100 individual names</span>
</span><span class='line'><span class="c1">## @loc.names: vector of  1 locus names</span>
</span><span class='line'><span class="c1">## @loc.nall: number of alleles per locus</span>
</span><span class='line'><span class="c1">## @loc.fac: locus factor for the  15 columns of @tab</span>
</span><span class='line'><span class="c1">## @all.names: list of  1 components yielding allele names for each locus</span>
</span><span class='line'><span class="c1">## @ploidy:  2</span>
</span><span class='line'><span class="c1">## @type:  codom</span>
</span><span class='line'><span class="c1">## </span>
</span><span class='line'><span class="c1">## Optionnal contents: </span>
</span><span class='line'><span class="c1">## @pop:  factor giving the population of each individual</span>
</span><span class='line'><span class="c1">## @pop.names:  factor giving the population of each individual</span>
</span><span class='line'><span class="c1">## </span>
</span><span class='line'><span class="c1">## @other: - empty -</span>
</span><span class='line'><span class="c1">## </span>
</span></code></pre></td></tr></table></div></figure>


<p>Once you&#8217;ve converted a sequence file into a <code>genind</code> object it behaves
just as you&#8217;d expect:</p>

<figure class='code'><figcaption><span></span></figcaption><div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='r'><span class='line'>diff_stats<span class="p">(</span>fake_genind<span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>




<figure class='code'><figcaption><span></span></figcaption><div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
</pre></td><td class='code'><pre><code class='r'><span class='line'><span class="c1">## $per.locus</span>
</span><span class='line'><span class="c1">##        Hs     Ht      Gst Gprime_st      D</span>
</span><span class='line'><span class="c1">## L1 0.9257 0.9332 0.008113    0.2165 0.2037</span>
</span><span class='line'><span class="c1">## </span>
</span><span class='line'><span class="c1">## $global</span>
</span><span class='line'><span class="c1">##        Hs        Ht   Gst_est Gprime_st     D_het    D_mean </span>
</span><span class='line'><span class="c1">##  0.925657  0.933228  0.008113  0.216513  0.203696  0.203696 </span>
</span><span class='line'><span class="c1">## </span>
</span></code></pre></td></tr></table></div></figure>



</div>
  
  


    </article>
  
  
    <article>
      
  <header>
    
      <h1 class="entry-title"><a href="/blog/2012/08/14/rentrez-an-r-package-for-interfacing-with-the-ncbis-databases/">Rentrez - an R Package for Interfacing With the NCBI&#8217;s Databases</a></h1>
    
    
      <p class="meta">
        








  


<time datetime="2012-08-14T18:59:00+12:00" pubdate data-updated="true">Aug 14<span>th</span>, 2012</time>
        
      </p>
    
  </header>


  <div class="entry-content"><p>Biologists make a lot of data these days. In fact, we seem to make more data
than we can really deal with. Databases like <a href="http://www.pubmed.org">Pubmed</a>,
<a href="http://www.fishbase.org">Fishbase</a>, <a href="http://www.gbif.org">GBIF</a> and the <a href="http://www.ncbi.nlm.nih.gov/sites/gquery">rest of the NCBI&#8217;s
offerings</a> are full of answers to
questions that no one has got around to asking.</p>

<p><a href="http://ropensci.org/">ROpenSci</a> is a collection of libraries for the R progamming
language that is designed to help biologists pull data from all over the web
into their R sessions and speed the process of turning data into information. The
<code>ROpenSci</code> team have already made a great set of tools, which allow researchers to
work with <a href="http://ropensci.org/packages/index.html">a bunch of databases</a>, and,
just as importantly share their results as nice reproducable R code. Up until
recently R has not really had a library for taking data from the NCBI&#8217;s various
databses (including such massive sources of data as Genbank and Pubmed). I&#8217;m very
happy to say that my library for doing just this,
 <a href="https://github.com/ropensci/rentrez"><code>rentrez</code></a> is now part of the <code>ROpenSci</code> family.</p>

<h2>What rentrez does</h2>

<p>At present the functions provided by <code>rentrez</code> cover the entire
<a href="http://www.ncbi.nlm.nih.gov/books/NBK25501/">Eutils API</a>. Basically, the functions
take arguments provided by a user, produce the URL needed to query the NCBI&#8217;s API
and fetches the resulting data. In most cases the functions return lists that
contain the parts of the resulting file that are most likely to be useful as
items. When the returned file is XML the list contains the XML file for those
that want to dig deeper.</p>

<h2>What you can do</h2>

<p>Grab a fork! The idea of ROpenSci is to create a set of tools that are useful to
as many scientists as possble. That means open data and open code.
If you find a bug in what I&#8217;ve done I want to know about it, if you think it can
be usefully extended feel free to pick it up and run. I&#8217;m also keen to here of
use-cases other than the ones already outlined in the documentation.</p>

<h2>An example usage</h2>

<p>There are some short examples in the README file for the <code>rentrez</code> repository,
but I thought I&#8217;d run through a longer one here. This is also reproducable blogging,
the markdown underlying this post was processed with <code>knitr</code> from <a href="https://gist.github.com/820f1fe0ed42ec7b6a15">this source</a></p>

<p>Lately, I&#8217;ve been working on a little meta-analysis of phylogenies. In particualr,
we&#8217;re interested in why sometimes different genes tell different stories about
the relationships between species from which the come. In terms of being able
to get the individual gene trees I need to do these analyses there are good,
rather less good and quite bad papers out there. In the best cases I can just
download the trees as nice, parsable newick files from<a href="treebase.org">TreeBase</a>
(which has already been <a href="github.com/ROpenSci/treebase">wrapped by ROpenSci</a>).
Sometimes I need to print out the trees from a paper and work with pencil and paper,
which I can handle. In a few cases people haven&#8217;t actually published their individual
gene trees, if I want to included these papers I need to replicate their work by
downloading the gene sequences, aligning them and making new trees.</p>

<p>So, here&#8217;s an example of how I&#8217;ve been using <code>rentrez</code> to automate some of that
process. I&#8217;m going to use a slightly convaluted process to get all the data, but
that&#8217;s just so I can walk though a bunch of the <code>rentrez</code> functions. Let&#8217;s get
started. Reece et al (2010, <a href="http://dx.doi.org/10.1016/j.ympev.2010.07.013">doi:10.1016/j.ympev.2010.07.013</a>)
presented a phylogeny of moray eels using four different genes, but didn&#8217;t
publish the gene trees. I want to get the sequences underlying their analyses,
which will be in the NCBI&#8217;s databases, so I can reproduce their results. To get
data associated with this paper from the NCBI I need the PMID (pubmed ID), which
I can find using the <code>rentrez</code> function <code>entrez_search</code> to query the pubmed
database with the paper&#8217;s doi:</p>

<figure class='code'><figcaption><span></span></figcaption><div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
</pre></td><td class='code'><pre><code class='r'><span class='line'>library<span class="p">(</span>rentrez<span class="p">)</span>
</span><span class='line'>pubmed_search <span class="o">&lt;-</span> entrez_search<span class="p">(</span>db <span class="o">=</span> <span class="s">&quot;pubmed&quot;</span><span class="p">,</span> term <span class="o">=</span> <span class="s">&quot;10.1016/j.ympev.2010.07.013[doi]&quot;</span><span class="p">)</span>
</span><span class='line'>pubmed_search<span class="p">$</span>ids
</span></code></pre></td></tr></table></div></figure>




<figure class='code'><figcaption><span></span></figcaption><div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
</pre></td><td class='code'><pre><code class='r'><span class='line'><span class="c1">## [1] 20674752</span>
</span></code></pre></td></tr></table></div></figure>


<p>All the functions in <code>rentrez</code> create a URL to get data from the NCBI, then fetch
the resulting document, usually as an XML file. In most cases the functions will
parse the most relevant sections of the XML file out and present them to you
as items in a list (<code>ids</code> being one item of the <code>pubmed_search</code> list in this case).</p>

<p>OK, now we have the PMID, what data does NCBI have for this paper? The
<code>entrez_link</code> function lets us find out. In this case the <code>db</code> argument can be
used to limit the number of data sources to check, but I want to see every data
source here so I&#8217;ll set this paramater to &#8220;all&#8221;:</p>

<figure class='code'><figcaption><span></span></figcaption><div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='r'><span class='line'>NCBI_data <span class="o">&lt;-</span> entrez_link<span class="p">(</span>dbfrom <span class="o">=</span> <span class="s">&quot;pubmed&quot;</span><span class="p">,</span> id <span class="o">=</span> pubmed_search<span class="p">$</span>ids<span class="p">,</span> db <span class="o">=</span> <span class="s">&quot;all&quot;</span><span class="p">)</span>
</span><span class='line'>str<span class="p">(</span>NCBI_data<span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>




<figure class='code'><figcaption><span></span></figcaption><div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
</pre></td><td class='code'><pre><code class='r'><span class='line'><span class="c1">## List of 10</span>
</span><span class='line'><span class="c1">##  $ pubmed_nuccore            : chr [1:119] &quot;307082467&quot; &quot;307082465&quot; &quot;307082463&quot; &quot;307082461&quot; ...</span>
</span><span class='line'><span class="c1">##  $ pubmed_popset             : chr [1:4] &quot;307082412&quot; &quot;307075396&quot; &quot;307075338&quot; &quot;307075274&quot;</span>
</span><span class='line'><span class="c1">##  $ pubmed_protein            : chr [1:118] &quot;307082468&quot; &quot;307082466&quot; &quot;307082464&quot; &quot;307082462&quot; ...</span>
</span><span class='line'><span class="c1">##  $ pubmed_pubmed             : chr [1:126] &quot;20674752&quot; &quot;20375076&quot; &quot;19053846&quot; &quot;11430656&quot; ...</span>
</span><span class='line'><span class="c1">##  $ pubmed_pubmed_combined    : chr [1:6] &quot;20674752&quot; &quot;20375076&quot; &quot;19053846&quot; &quot;11430656&quot; ...</span>
</span><span class='line'><span class="c1">##  $ pubmed_pubmed_five        : chr [1:6] &quot;20674752&quot; &quot;20375076&quot; &quot;19053846&quot; &quot;11430656&quot; ...</span>
</span><span class='line'><span class="c1">##  $ pubmed_pubmed_reviews     : chr &quot;20674752&quot;</span>
</span><span class='line'><span class="c1">##  $ pubmed_pubmed_reviews_five: chr &quot;20674752&quot;</span>
</span><span class='line'><span class="c1">##  $ pubmed_taxonomy_entrez    : chr [1:40] &quot;876649&quot; &quot;876647&quot; &quot;876643&quot; &quot;876642&quot; ...</span>
</span><span class='line'><span class="c1">##  $ file                      :Classes &#39;XMLInternalDocument&#39;, &#39;XMLAbstractDocument&#39;, &#39;oldClass&#39; &lt;externalptr&gt; </span>
</span></code></pre></td></tr></table></div></figure>


<p>The most relevant data here is the from the <a href="ncbi.nlm.nih.gov/popset">popset</a>
database, which containts population and phylogenetic datasets. If I want to
see what each of the four popset datasets associated with this paper are about I
can use <code>entrez_summary</code> to have a look. This function can collect summaries
from a lot of different databases, and, because the XML return by those databases
isn&#8217;t conisitant doesn&#8217;t make any attempt to parse information from the resulting
file. Instead you get a <code>XMLInternalDocument</code> object from the <code>XML</code> library, which
you have to further process yourself. In this case, a little xpath gets the name
of each dataset:</p>

<figure class='code'><figcaption><span></span></figcaption><div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
</pre></td><td class='code'><pre><code class='r'><span class='line'>data_summaries <span class="o">&lt;-</span> entrez_summary<span class="p">(</span>db <span class="o">=</span> <span class="s">&quot;popset&quot;</span><span class="p">,</span> ids <span class="o">=</span> NCBI_data<span class="p">$</span>pubmed_popset<span class="p">)</span>
</span><span class='line'>xpathSApply<span class="p">(</span>data_summaries<span class="p">,</span> <span class="s">&quot;//Item[@Name=&#39;Title&#39;]&quot;</span><span class="p">,</span> xmlValue<span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>




<figure class='code'><figcaption><span></span></figcaption><div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
</pre></td><td class='code'><pre><code class='r'><span class='line'><span class="c1">## [1] &quot;Muraenidae cytochrome oxidase subunit 1 gene, partial cds; mitochondrial.&quot;</span>
</span><span class='line'><span class="c1">## [2] &quot;Muraenidae recombination activating protein 2 gene, partial cds.&quot;         </span>
</span><span class='line'><span class="c1">## [3] &quot;Muraenidae recombination activating protein 1 gene, partial cds.&quot;         </span>
</span><span class='line'><span class="c1">## [4] &quot;Muraenidae cytochrome b gene, partial cds; mitochondrial.&quot;                </span>
</span></code></pre></td></tr></table></div></figure>


<p>Ok, since we might expect nuclear and mitochondrial genes to hav different
histories, let&#8217;s get sequences from each genome (the the COI and RAG1 datasets)
using <code>entrez_fetch</code>. By specifying <code>file_format="fasta"</code> we will get characater
vectors in the fasta format:</p>

<figure class='code'><figcaption><span></span></figcaption><div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
</pre></td><td class='code'><pre><code class='r'><span class='line'>coi <span class="o">&lt;-</span> entrez_fetch<span class="p">(</span>db <span class="o">=</span> <span class="s">&quot;popset&quot;</span><span class="p">,</span> ids <span class="o">=</span> NCBI_data<span class="p">$</span>pubmed_popset<span class="p">[</span><span class="m">1</span><span class="p">],</span> file_format <span class="o">=</span> <span class="s">&quot;fasta&quot;</span><span class="p">)</span>
</span><span class='line'>rag1 <span class="o">&lt;-</span> entrez_fetch<span class="p">(</span>db <span class="o">=</span> <span class="s">&quot;popset&quot;</span><span class="p">,</span> ids <span class="o">=</span> NCBI_data<span class="p">$</span>pubmed_popset<span class="p">[</span><span class="m">3</span><span class="p">],</span> file_format <span class="o">=</span> <span class="s">&quot;fasta&quot;</span><span class="p">)</span>
</span><span class='line'>write<span class="p">(</span>coi<span class="p">,</span> <span class="s">&quot;moray_coi_raw.fasta&quot;</span><span class="p">)</span>
</span><span class='line'>write<span class="p">(</span>rag1<span class="p">,</span> <span class="s">&quot;moray_rag1_raw.fasta&quot;</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p>So I&#8217;ve got the data on hand - that&#8217;s all the I need <code>rentrez</code> for, but I might
as well align these  sequences and make gene trees for each. I&#8217;ll just do a
quick and diry neighbor-joining tree using <code>ape</code> and we can clean up the long
OTU names with the help of <code>stingr</code>. (I put the fussy work of cleaning the names
 and rooting the trees into a function <code>clean_and_root</code>):</p>

<figure class='code'><figcaption><span></span></figcaption><div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class='line-number'>1</span>
<span class='line-number'>2</span>
<span class='line-number'>3</span>
<span class='line-number'>4</span>
<span class='line-number'>5</span>
<span class='line-number'>6</span>
<span class='line-number'>7</span>
<span class='line-number'>8</span>
<span class='line-number'>9</span>
<span class='line-number'>10</span>
<span class='line-number'>11</span>
<span class='line-number'>12</span>
<span class='line-number'>13</span>
<span class='line-number'>14</span>
<span class='line-number'>15</span>
<span class='line-number'>16</span>
</pre></td><td class='code'><pre><code class='r'><span class='line'>library<span class="p">(</span>ape<span class="p">)</span>
</span><span class='line'>library<span class="p">(</span>stringr<span class="p">)</span>
</span><span class='line'>clean_and_root <span class="o">&lt;-</span> <span class="kr">function</span><span class="p">(</span>tr<span class="p">,</span> outgroup<span class="p">,</span> resolved <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span> <span class="p">{</span>
</span><span class='line'>    tr<span class="p">$</span>tip.label <span class="o">&lt;-</span> sapply<span class="p">(</span>str_split<span class="p">(</span>tr<span class="p">$</span>tip.label<span class="p">,</span> <span class="s">&quot; &quot;</span><span class="p">),</span> <span class="kr">function</span><span class="p">(</span>x<span class="p">)</span> paste<span class="p">(</span>x<span class="p">[</span><span class="m">2</span>:<span class="m">3</span><span class="p">],</span>
</span><span class='line'>        collapse <span class="o">=</span> <span class="s">&quot;_&quot;</span><span class="p">))</span>
</span><span class='line'>    <span class="kr">return</span><span class="p">(</span>root<span class="p">(</span>tr<span class="p">,</span> outgroup<span class="p">,</span> resolve.root <span class="o">=</span> resolved<span class="p">))</span>
</span><span class='line'><span class="p">}</span>
</span><span class='line'>par<span class="p">(</span>mfrow <span class="o">=</span> c<span class="p">(</span><span class="m">1</span><span class="p">,</span> <span class="m">2</span><span class="p">))</span>
</span><span class='line'>coi_ali <span class="o">&lt;-</span> muscle<span class="p">(</span>read.dna<span class="p">(</span><span class="s">&quot;moray_coi_raw.fasta&quot;</span><span class="p">,</span> <span class="s">&quot;fasta&quot;</span><span class="p">))</span>
</span><span class='line'>coi_tr <span class="o">&lt;-</span> nj<span class="p">(</span>dist.dna<span class="p">(</span>coi_ali<span class="p">,</span> <span class="s">&quot;k81&quot;</span><span class="p">))</span>
</span><span class='line'>clean_coi_tr <span class="o">&lt;-</span> clean_and_root<span class="p">(</span>coi_tr<span class="p">,</span> <span class="s">&quot;Uropterygius_macrocephalus&quot;</span><span class="p">)</span>
</span><span class='line'>plot<span class="p">(</span>clean_coi_tr<span class="p">,</span> direction <span class="o">=</span> <span class="s">&quot;rightwards&quot;</span><span class="p">,</span> cex <span class="o">=</span> <span class="m">0.5</span><span class="p">)</span>
</span><span class='line'>rag_ali <span class="o">&lt;-</span> muscle<span class="p">(</span>read.dna<span class="p">(</span><span class="s">&quot;moray_rag1_raw.fasta&quot;</span><span class="p">,</span> <span class="s">&quot;fasta&quot;</span><span class="p">))</span>
</span><span class='line'>rag_tr <span class="o">&lt;-</span> nj<span class="p">(</span>dist.dna<span class="p">(</span>rag_ali<span class="p">,</span> <span class="s">&quot;k81&quot;</span><span class="p">))</span>
</span><span class='line'>clean_rag_tr <span class="o">&lt;-</span> clean_and_root<span class="p">(</span>rag_tr<span class="p">,</span> <span class="s">&quot;Uropterygius_macrocephalus&quot;</span><span class="p">)</span>
</span><span class='line'>plot<span class="p">(</span>clean_rag_tr<span class="p">,</span> direction <span class="o">=</span> <span class="s">&quot;leftward&quot;</span><span class="p">,</span> cex <span class="o">=</span> <span class="m">0.5</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure>


<p><img src="http://i.imgur.com/QcghP.png" alt="plot of chunk trees" /></p>
</div>
  
  


    </article>
  
  <div class="pagination">
    
    <a href="/blog/archives">Blog Archives</a>
    
  </div>
</div>
<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2012/11/17/diff-stats-for-sequence-data/">MMOD does sequences</a>
      </li>
    
      <li class="post">
        <a href="/blog/2012/08/14/rentrez-an-r-package-for-interfacing-with-the-ncbis-databases/">rentrez - an R package for interfacing with the NCBI&#8217;s databases</a>
      </li>
    
  </ul>
</section>

<section>
  <h1>GitHub Repos</h1>
  <ul id="gh_repos">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  
  <a href="https://github.com/dwinter">@dwinter</a> on GitHub
  
  <script type="text/javascript">
    $.domReady(function(){
        if (!window.jXHR){
            var jxhr = document.createElement('script');
            jxhr.type = 'text/javascript';
            jxhr.src = '/javascripts/libs/jXHR.js';
            var s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(jxhr, s);
        }

        github.showRepos({
            user: 'dwinter',
            count: 0,
            skip_forks: true,
            target: '#gh_repos'
        });
    });
  </script>
  <script src="/javascripts/github.js" type="text/javascript"> </script>
</section>


<section>
  <h1>Latest Tweets</h1>
  <ul id="tweets">
    <li class="loading">Status updating&#8230;</li>
  </ul>
  <script type="text/javascript">
    $.domReady(function(){
      getTwitterFeed("theatavism", 3, false);
    });
  </script>
  <script src="/javascripts/twitter.js" type="text/javascript"> </script>
  
    <a href="http://twitter.com/theatavism" class="twitter-follow-button" data-show-count="false">Follow @theatavism</a>
  
</section>





  
</aside>

    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2012 - David Winter -
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a></span>
</p>

</footer>
  







  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
